<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Иван! Давай знакомиться! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы». 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
    
Все ключевые этапы в работе выполнены, и я вижу что с проектом ты справшяешься. Есть моменты, которые нужно доработать, но я уверен, у тебя все получится.
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂
 
<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>образец комментария студента</b></font>
   
Можешь использовать этот комментарий. Просто удали этот текст и вместо него напиши свой 🙂

</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-подготовке" data-toc-modified-id="Вывод-по-подготовке-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод по подготовке</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-подготовке" data-toc-modified-id="Вывод-по-подготовке-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод по подготовке</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

In [1]:
import pandas as pd
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
from nltk.corpus import wordnet
from nltk import pos_tag
from tqdm.notebook import tqdm
import numpy as np
import nltk
import re
RANDOM_STATE = 42

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Но вот лишние импоры лучше убрать, поберечь ресурсы</div>

## Подготовка

In [2]:
try:
    data = pd.read_csv('toxic_comments.csv')
except:
    data = pd.read_csv('/datasets/toxic_comments.csv')

In [3]:
data.head(5)

Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

После загрузки мы видим 3 столбца: неизвестный, текст, показатель токсичности

In [4]:
data.isna().sum()

Unnamed: 0    0
text          0
toxic         0
dtype: int64

Пропусков в данных не обнаружено

In [5]:
data.duplicated().sum()

0

Дубликатов в данных не обнаружено

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Данные загружены и исследованы.</div>

In [7]:
data = data.drop("Unnamed: 0", axis = 1)
data.head(5)

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

Выполним лемматизацию текста и удаление ненужных символов

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    
Да, все верно, столбец `'Unnamed: 0'` можно безболезненно удалить. Это "старый" индекс, который похоже забыли при пересохранении датасета.</div>

In [8]:
corpus = data['text'].values


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
В нашем проекте не нужно (и даже вредно) менять кодировку на unicode. Изменение кодировки увеличивает количество оперативной памяти, необходимой для хранения текстов и может привсти к падению ядра. 
    
В теоретической части мы работали с русскоязычными текстами, поэтому изменение кодировки было обосновано. Но тексты в проекте  написаны латиницей, поэтому вполне достаточно кодировки по умолчанию.</div>


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> исправлено 👍 </div>

In [9]:
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /home/jovyan/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/jovyan/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [10]:
lemmatizer = WordNetLemmatizer()

def get_wordnet_pos(tag):
    if tag.startswith('J'):
        return wordnet.ADJ
    elif tag.startswith('V'):
        return wordnet.VERB
    elif tag.startswith('N'):
        return wordnet.NOUN
    elif tag.startswith('R'):
        return wordnet.ADV
    else:
        return wordnet.NOUN

def lemm_and_clear(text):
    tokens = word_tokenize(text)
    tagged_tokens = pos_tag(tokens)
    lemmas = [lemmatizer.lemmatize(token, get_wordnet_pos(tag)) for token, tag in tagged_tokens]
    lemmatized_text = ' '.join(lemmas)
    cleaned_text = re.sub(r'[^a-zA-Z\' ]', '', lemmatized_text)
    cleaned_text = cleaned_text.replace('\n', '')
    
    return cleaned_text

corpus = [lemm_and_clear(text) for text in corpus]

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Верно использован WordNetLemmatizer</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">

Молодец, что умеешь list comprehatnion, но в нашем случае лучше отказаться от циклов, даже в такой форме.
    
Гораздо более читаемым выглядит использование apply, и хоть "под капотом" там тоже циклы, насколько я понимаю там ищется возможность для оптимизации через векторизацию (используется numba).
    
    
---------------------------------
    
Правильная лемматизация длится долго... Минут 20-30, в зависимости от мощности компьютера. Гораздо комфортнее видеть прогресс выполнения этой длительной операции, чем сидеть и гадать "а не зависла ли она", "закончит только к утру или через 30 секунд"? Можно воспользоваться прогресс -баром от tqdm.
    
     
    from tqdm.notebook import tqdm
    tqdm.pandas()

    data['lemm_text'] = data['text'].progress_apply(lemmafunction)
        
    
    
</div>



После длительной лемматизации текста получаем необходимый набор текстов

In [11]:
data['text'] = corpus

После лемматизации могли появиться дубликаты, потому удаляем их

In [12]:
data = data.drop_duplicates(subset='text')

In [13]:
stop_words = set(stopwords.words('english'))

Выполним разбивку текста на тренировочную и тестовую выборки

In [14]:
X = data.drop("toxic", axis = 1)
Y = data['toxic']
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.4, random_state = RANDOM_STATE)
X_train.head(5)

text
46180    I be give advice   '' You may wish to conside...
115283   photo in  Baghdad Church Attack article Lihaa...
60948   No problem  Ylee  happen to me all the time  I...
67085             Wow  way to keep pull StrawMen out o...
7934     Please do not vandalize page  a you do with t...

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    
Данные разделены на выборки. Но я бы посоветовал тебе изменить пропорции и выделить на тест 10%. Причина простая - модели, которые мы обучаем чувствительны к объему обучающих данных. Чем больше слов они увидят в процессе обучения, и оценят их вклад в "токсичность", тем лучше будут модели. А для корректного тестирования и 10% данных вполне достаточно, учитывая немаленький размер датасета.

</div>

In [16]:
X_train

text
46180    I be give advice   '' You may wish to conside...
115283   photo in  Baghdad Church Attack article Lihaa...
60948   No problem  Ylee  happen to me all the time  I...
67085             Wow  way to keep pull StrawMen out o...
7934     Please do not vandalize page  a you do with t...
...                                                   ...
120229   I HAVE TO WAIT  MORE FUCKING DAYS  I 'm sick ...
103971  Actually you know what  I 'm not use this acco...
132350  Welcome  Hello   and welcome to Wikipedia  Tha...
147362  DYK nomination of LASNEX Hello  Your submissio...
122316    what be the specific problem here  and how c...

[95239 rows x 1 columns]

In [17]:
X_tr_l = list(X_train['text'])
X_te_l = list(X_test['text'])

Выполним трансформацию текстов в векторы при попомщи tf-idf

In [18]:
count = TfidfVectorizer(stop_words = stop_words)
tf_idf_tr = count.fit_transform(X_tr_l)
tf_idf_te = count.transform(X_te_l)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
👍 Молодец, получены признаки из текстов с помощью TF-IDF. И совершенно верно векторайзер вначале обучен на обучающем наборе и затем применен к тестовому</div>

### Вывод по подготовке
- Не выявлено дубликатов и пропущенных значений перед лемматизацией
- Выполнена лемматизация
- Устранены дубликаты, возникшие после лемматизации

## Обучение

Обучим 2 модели с перебором гиперапараметров, а именно:
- Логистическую регрессию с l1, l2 регуляризациями
- Дерево решений со значением глубины

In [19]:
model_log_reg = LogisticRegression(solver='liblinear', random_state = RANDOM_STATE)

param_grid = {
    'penalty': ['l1', 'l2'],
    'C': list(range(1, 15, 3))
}

grid_search_lr = GridSearchCV(
    model_log_reg, 
    param_grid, 
    cv=5, 
    scoring='f1'
)

grid_search_lr.fit(tf_idf_tr, y_train)

print("Best parameters:", grid_search_lr.best_params_)
print("Best F1 score:", grid_search_lr.best_score_)

Best parameters: {'C': 4, 'penalty': 'l1'}
Best F1 score: 0.7656120123750203


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Поиск по сетке использует крос-валидацию.  А кросс-валидация одновремено с извлечением признаков из текстов требует особого подхода. Ты помнишь, что векторайзер мы обучали на обучающей выборке а применяли к тестовой. Что происходит при кросс-валидации? Ты разбиваешь большую обучающую выборку на несколько фолдов, из них выделяешь часть на обучение, часть на валидацию.  Но признаки (TF-IDF) уже обучены на всем наборе (обучающий + валидационный). Получается утечка данных. Чтобы избежать её нужно использовать pipeline из векторайзера и модели.
    
Пример такого пайплайна можно посмотреть в официальной документации
    
  https://scikit-learn.org/stable/auto_examples/model_selection/plot_grid_search_text_feature_extraction.html
    
И еще полезные материалы:
    
  https://newtechaudit.ru/pipeline-v-rabote-s-dannymi/      
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
А вот подбор гиперпараметра С и работа с регуляризацией это очень правильное решение. Линейные модели очень хорошо работают с признаками, полученными из текстов. С помощью TF-IDF мы получили очень длинные разряженные вектора. И действительно, они очень длинные (длина около 140 000), при этом очень мало значений отличаются от нуля. Так вот, линейные модели гораздо лучше деревьев справляются с такими признаками. Но им нужно помочь - и ослабить регуляризацию, т.к. она заметно вырастает из за большого количества признаков. И делаем мы это как раз подбирая гиперпараметр С.
    

</div>

На тренировочной выборке получена лучшая модель с l1 регуляризацией и коэффициентом C = 4, f1 = 0.766

In [23]:
model_dt = DecisionTreeClassifier(random_state=RANDOM_STATE)

param_grid = {
    'max_depth': [4, 6, 8]
}

grid_search_dt = GridSearchCV(
    model_dt, 
    param_grid, 
    cv=5, 
    scoring='f1'
)

grid_search_dt.fit(tf_idf_tr, y_train)

print("Best parameters:", grid_search_dt.best_params_)
print("Best F1 score:", grid_search_dt.best_score_)

Best parameters: {'max_depth': 8}
Best F1 score: 0.5613298417159022


На тренировочной выборке получена лучшая модель с глубиной = 8, f1 = 0.56

Дерево решений показало себя значительно хуже чем Лог регрессия, потому предикт будем делать именно на ней

In [24]:
preds = grid_search_lr.predict(tf_idf_te)
print(f1_score(y_test, preds))

0.7756209205730271


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец: 
    
  - выбрана лучшая модель
  - тестирование произведено корректно 
  - достигнута метрика f1 выше 0,75 на тестовой выборке
</div>


Значение F1-метрики превышает пороговое значение заказчика, потому мы подобрали хорошую модель, удовлетворяющую требованиям, а именно:
- LogisticRegression {'C': 4, 'penalty': 'l1'}

## Выводы

В данном проекте я работал сервисом, подобным википедии, где мне пришлось разработать модель, классифицирующую правки пользователей на токсичность. Итоговая модель (LogisticRegression {'C': 4, 'penalty': 'l1'}) имеет значение F1-метрики, превышающее пороговое значение заказчика (F1 = 0.7756), а значит данную модель можно использовать заказчику.
А также перед обучением была выполнена предобработка данных:

### Вывод по подготовке
- Не выявлено дубликатов и пропущенных значений перед лемматизацией
- Выполнена лемматизация
- Устранены дубликаты, возникшие после лемматизации

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий v2</b>

Проект принят. Поздравляю и желаю дальнейших успехов!
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий</b>
Давай подведем итоги. В целом с проектом ты справляешься - текст предобработан, извлечены признаки и обучены классификаторы. Достигнуто требуемое значение метрики f1.
    
Единственный красный комментарий связан с кодировками: Не стоит перекодировать английский текст в Юникод. Уверен, исправления будут легкими и быстрыми ))
 
     
Жду твоих исправлений :)
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> рекомендации по доп. материалам</b>
Если решишь погрузиться в область работы с текстами, очень советую несколько продвинутых бесплатных курсов.
    
   - Отличный бесплатный курс от Школы глубокого обучения МФТИ (https://stepik.org/org/dlschool), старт курса каждые пол года. Два семестра, один по основам и компьютерному зрению, второй по обработке естественного языка. Проходить нужно именно в таком порядке,т.к. почти весь современный NLP построен на нейронках.
    
   - "Нейронные сети и компьютерное зрение" от Samsung Research Russia (https://stepik.org/course/50352/syllabus). Есть также продолжение по NLP.  
   - Трек NLP от сообщества ODS https://ods.ai/tracks/nlp-course-autumn-22
    
    
А если на тебя произвели впечатление возможности ChatGPT и хочешь попробовать использовать возможности больших языковых моделей для решения своих задач, могу порекомендовать следующие курсы (но к сожалению все на английском).
    
  - https://www.coursera.org/learn/generative-ai-with-llms  (можно прослужать бесплатно)
  - Короткие курсы на сайте https://www.deeplearning.ai/short-courses/  Самый свежак - как писать промпты, как использовать LLM для создания собственных приложений (например как создать бота на основе ChatGPT, который будет отвечать на вопросы по внутренней документации компании).  
  - https://www.promptingguide.ai/introduction/settings
    
    
Также можно обратиться к туториалу по использованию языковых моделей на HugingFace. Обрати внимание, для больше чем половины материалов доступен перевод на русский, хотя можно читать и на языке оригинала. 
    
  https://huggingface.co/learn/nlp-course   
 
</div>